tower-hedge/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+#[macro_use]
+extern crate futures;
+extern crate hdrhistogram;
+#[macro_use]
+extern crate log;
+extern crate tokio_timer;
+extern crate tower_service;
+
+use futures::{Async, Future, Poll};
+use hdrhistogram::Histogram;
+use tokio_timer::{clock, Delay};
+use tower_service::Service;
+
+use std::sync::{Arc, Mutex};
+use std::time::{Duration, Instant};
+
+mod rotating;
+
+use rotating::Rotating;
+
+/// A "retry policy" to classify if a request should be pre-emptively retried.
+pub trait Policy<Request>: Sized {
+    /// Check the policy if a certain request should be pre-emptively retried.
+    ///
+    /// This method is passed a reference to the original request.
+    fn can_retry(&self, req: &Request) -> bool;
+    /// Tries to clone a request before being passed to the inner service.
+    ///
+    /// If the request cannot be cloned, return `None`.
+    fn clone_request(&self, req: &Request) -> Option<Request>;
```

Comment:
It looks like `clone_request` is called at the beginning, and then `can_retry` is called after the delay triggers. Do you envision use cases where the knowledge of whether a certain request should be retried would only be known later? Or would that knowledge be generally known when `clone_request` is being called?

Reply:
For example, retry budgets should be checked after the delay has triggered, when the hedge request is about to be issued.